File: virtual-cube/src/draw/BezierFacelet.java
package draw;

import java.awt.Color;
import java.awt.Graphics2D;
import java.awt.geom.GeneralPath;

public class BezierFacelet extends Facelet {

	public int n;
	public Vertex[] vertex;

	public BezierFacelet(int n) {
		this(n, Color.BLACK, Color.BLACK);
	}

	public BezierFacelet(int n, Color color_bound, Color color_inside) {
		this.n = n;
		vertex = new Vertex[3 * n];
		this.color_bound = color_bound;
		this.color_inside = color_inside;
	}

	public BezierFacelet(BezierFacelet p) {
		this.n = p.n;
		this.color_bound = p.color_bound;
		this.color_inside = p.color_inside;
		this.center = new Vertex(p.center);
		this.vertex = new Vertex[3 * p.n];
		for (int i = 0; i < 3 * p.n; ++i) {
			this.vertex[i] = new Vertex(p.vertex[i]);
		}
	}

	public void rotate(Vertex axis, double angle) {
		center.rotate(axis, angle);
		for (int i = 0; i < 3 * n; ++i) {
			vertex[i].rotate(axis, angle);
		}
	}

	public void draw(Graphics2D g, Vertex eye, Vertex axis_x, Vertex axis_y,
			double dx, double dy, double ratio) {
		double[][] vertex2d = new double[3 * n][2];
		for (int i = 0; i < 3 * n; ++i) {
			vertex2d[i] = vertex[i].project(eye, axis_x, axis_y);
			vertex2d[i][0] = vertex2d[i][0] * ratio + dx;
			vertex2d[i][1] = vertex2d[i][1] * ratio + dy;
		}
		GeneralPath path = new GeneralPath();
		path.moveTo(vertex2d[0][0], vertex2d[0][1]);
		for (int i = n; --i >= 0;) {
			path.curveTo(vertex2d[3 * i + 2][0], vertex2d[3 * i + 2][1],
					vertex2d[3 * i + 1][0], vertex2d[3 * i + 1][1],
					vertex2d[3 * i][0], vertex2d[3 * i][1]);
		}
		Color current_color = g.getColor();
		g.setColor(color_inside);
		g.fill(path);
		g.setColor(color_bound);
		g.draw(path);
		g.setColor(current_color);
	}
}

File: canvas/bezier-facelet.js
var BezierFacelet = function(vertex, control, color_bound, color_inside) {
	this.vertex = vertex;
	this.control = control;
	this.color_bound = color_bound || null;
	this.color_inside = color_inside || null;
};

BezierFacelet.prototype.clone = function() {
	var facelet = new BezierFacelet(this.vertex, this.control,
			this.color_bound, this.color_inside);
	facelet.vertex = facelet.vertex.map(function(vertex) {
		return vertex.clone();
	});
	facelet.control = facelet.control.map(function(control) {
		return [control[0].clone(), control[1].clone()];
	});
	return facelet;
};

BezierFacelet.prototype.translate = function(x, y, z) {
	this.vertex.forEach(function(vertex) {
		vertex.translate(x, y, z);
	});
	this.control.forEach(function(control) {
		control[0].translate(x, y, z);
		control[1].translate(x, y, z);
	});
	return this;
};

BezierFacelet.prototype.scale = function(ratio_x, ratio_y, ratio_z) {
	this.vertex.forEach(function(vertex) {
		vertex.scale(ratio_x, ratio_y, ratio_z);
	});
	this.control.forEach(function(control) {
		control[0].scale(ratio_x, ratio_y, ratio_z);
		control[1].scale(ratio_x, ratio_y, ratio_z);
	});
	return this;
};

BezierFacelet.prototype.rotate = function(axis, angle) {
	this.vertex.forEach(function(vertex) {
		vertex.rotate(axis, angle);
	});
	this.control.forEach(function(control) {
		control[0].rotate(axis, angle);
		control[1].rotate(axis, angle);
	});
	return this;
};

BezierFacelet.prototype.skew = function(direction, ratio) {
	this.vertex.forEach(function(vertex) {
		vertex.skew(direction, ratio);
	});
	this.control.forEach(function(control) {
		control[0].skew(direction, ratio);
		control[1].skew(direction, ratio);
	});
	return this;
};

BezierFacelet.prototype.transform = function(matrix) {
	this.vertex.forEach(function(vertex) {
		vertex.transform(matrix);
	});
	this.control.forEach(function(control) {
		control[0].transform(matrix);
		control[1].transform(matrix);
	});
	return this;
};

BezierFacelet.prototype.draw = function(context, inspector, width) {
	var point = this.vertex.map(function(vertex) {
		return vertex.project(inspector);
	});
	point.push(point[0]);
	var control = this.control.map(function(pair) {
		return pair.map(function(point) {
			return point.project(inspector);
		});
	});
	context.beginPath();
	context.moveTo(point[0].x, point[0].y);
	for (var i = 0; i < control.length; ++i) {
		context.bezierCurveTo(
			control[i][0].x, control[i][0].y,
			control[i][1].x, control[i][1].y,
			point[i + 1].x, point[i + 1].y
		);
	}
	context.closePath();
	if (this.color_inside) {
		context.fillStyle = this.color_inside;
		context.fill();
	}
	if (this.color_bound && context.lineWidth) {
		context.lineWidth = width;
		context.strokeStyle = this.color_bound;
		context.stroke();
	}
};